packages/rum-core/test/common/context.spec.js

```diff
@@ -0,0 +1,170 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { addSpanContext, addTransactionContext } from '../../src/common/context'
+import resourceEntries from '../fixtures/resource-entries'
+import Span from '../../src/performance-monitoring/span'
+import Transaction from '../../src/performance-monitoring/transaction'
+import { PAGE_LOAD } from '../../src/common/constants'
+import { mockGetEntriesByType } from '../utils/globals-mock'
+
+describe('Context', () => {
+  function createSpanWithData(url) {
+    const span = new Span(`GET ${url}`, 'external')
+    const data = {
+      url,
+      method: 'GET',
+      response: {
+        status: 200
+      }
+    }
+    span.end()
+    addSpanContext(span, data)
+    return span
+  }
+
+  it('should add external span context with all fields', () => {
+    let url = 'http://user:pass@testing.local:1234/path?query'
+    let span = createSpanWithData(url)
+    expect(span.context).toEqual({
+      http: {
+        method: 'GET',
+        url: 'http://[REDACTED]:[REDACTED]@testing.local:1234/path?query',
+        status_code: 200
+      },
+      destination: {
+        service: {
+          name: 'http://testing.local:1234',
+          resource: 'testing.local:1234',
+          type: 'external'
+        },
+        address: 'testing.local',
+        port: '1234'
+      }
+    })
+
+    url = 'https://www.elastic.co:443/products/apm'
+    span = createSpanWithData(url)
+    expect(span.context).toEqual({
+      http: {
+        method: 'GET',
+        url,
+        status_code: 200
+      },
+      destination: {
+        service: {
+          name: 'https://www.elastic.co',
+          resource: 'www.elastic.co:443',
+          type: 'external'
+        },
+        address: 'www.elastic.co',
+        port: '443'
+      }
+    })
+
+    url = 'http://[::1]/'
+    span = createSpanWithData(url)
+    expect(span.context).toEqual({
+      http: {
+        method: 'GET',
+        url,
+        status_code: 200
+      },
+      destination: {
+        service: {
+          name: 'http://[::1]',
+          resource: '[::1]:80',
+          type: 'external'
+        },
+        address: '[::1]',
```

Comment:
Any reason why we need special treatment for address alone when compared to service.name and resource?

Reply:
Yes :) From https://github.com/elastic/apm/issues/115#issuecomment-562452269:

> Note: for IPv6 addresses, surrounding the address with square brackets is only relevant where you have (or may have) the port alongside, e.g. [::1]:80, http://[::1]:80, http://[::1] (square brackets still required because the port is optional). The square brackets are there to disambiguate the IPv6 address, which contains colons, from the proceeding port, which is separated by a colon.
>
> When recording an IPv6 address separately from a port, as we are in the case of context.destination.address, then it should be recorded in its canonical form: without square brackets, e.g. ::1.

In `destination.service.name` the port may is optionally present (so brackets are required either way to disambiguate), and in `destination.service.resource` the port is always present.

Reply:
Thanks for the explanation Andrew 👍